device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTpm.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Shared;
+using System;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// Authentication method that uses a shared access signature token and allows for token refresh. 
+    /// </summary>
+    public sealed class DeviceAuthenticationWithTpm : DeviceAuthenticationWithTokenRefresh
+    {
+        private SecurityClientHsmTpm _securityClient;
+        private int _timeToLiveSeconds;
+
+        public DeviceAuthenticationWithTpm(
+            string deviceId, 
+            SecurityClientHsmTpm securityClient,
+            int timeToLiveSeconds = 1 * 60 * 60,
+            int timeToLiveBufferSeconds = 10 * 60) : base(deviceId, timeToLiveBufferSeconds)
+        {
+            if (securityClient == null)
+            {
+                throw new ArgumentNullException(nameof(securityClient));
+            }
+
+            if (timeToLiveSeconds < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeToLiveSeconds));
+            }
+
+            if (timeToLiveBufferSeconds < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeToLiveBufferSeconds));
+            }
+
+            _securityClient = securityClient;
+            _timeToLiveSeconds = timeToLiveSeconds;
+        }
+        
+        protected override Task<string> SafeCreateNewToken(string iotHubName)
+        {
+            const long WINDOWS_TICKS_PER_SEC = 10000000;
+            const long EPOCH_DIFFERNECE = 11644473600;
```

Comment:
is this meant to be hardcoded or will it get updated?

Reply:
Hardcoded - Since we already have a token builder in the device client this will change with my next PR.